### PR TITLE
[Custom] Max Unpool custom layer implementation

### PR DESCRIPTION
**Pull request by dkurt**
### Details:
Proposal to add custom layer MaxUnpool into the Template CPU Extensions library. This is required for https://github.com/openvinotoolkit/openvino_notebooks/pull/46

TODO: MO extensions, how-to-use guide similar to FFT ext.?

### Usage
```python
import os
import openvino
from openvino.inference_engine import IECore

ie = IECore()
path = os.path.join(os.path.dirname(openvino.__file__), 'libs', 'libtemplate_extension.so')
ie.add_extension(path, 'CPU')
```

### CODE REVIEW

**Comment by ryanloney** [comment]
I just want to comment that this is needed for a partner product launch in July. Please merge this for 2021.4 release if possible.

**Comment by dkurt** [comment]
I'm a bit confused because this library is not compiled on CI at all. 

https://github.com/openvinotoolkit/openvino/blob/6a16b70e0ef7fba37052a2b7b3ddb63e8fc9b32e/docs/CMakeLists.txt#L5

**Comment by dkurt** [comment]
@mryzhov, do you know which instructions are responsible for the libraries filter for wheels packages? Is that possible to add `libtemplate_extension.so`?

**Comment by mryzhov** [comment]
> @mryzhov, do you know which instructions are responsible for the libraries filter for wheels packages? Is that possible to add `libtemplate_extension.so`?

This logic is implemented here - https://github.com/mryzhov/openvino/blob/271cf8146cf83ed2b747cde7c369413e0db8561b/inference-engine/ie_bridges/python/wheel/setup.py#L34

The simplest way is to include this lib to the existed CMake targets (for example core or cpu) then it will be automatically included in the wheel.

**Comment by dkurt** [comment]
@mryzhov, is that possible to download `.whl` from CI to verify?

**Comment by mryzhov** [comment]
> @mryzhov, is that possible to download `.whl` from CI to verify?

Yes, CI builds wheels for each commit/pre-commit

**Comment by dkurt** [comment]
@mryzhov, may I ask to review? Added a new install target but still no `libtemplate_extension.so` library in the package

**Comment by dkurt** [comment]
@slyubimt, @mryzhov, please guide me who can review this proposal? We need to add this functionality to `openvino-dev`. CI part is done so this is a final PR.

File: inference-engine/ie_bridges/python/wheel/setup.py [hunk 1]
```diff
@@ -87,6 +87,12 @@
         'install_dir': TBB_LIBS_DIR,
         'rpath': LIBS_RPATH,
     },
+    'cpu_extensions': {
+        'name': 'template_extension',
```
Comment (ilyachur) [inline inference-engine/ie_bridges/python/wheel/setup.py:91]: @dkurt Why do you add `template_extension` to python package? We shouldn't do it, `template_extension` is used only in OpenVINO documentation.

File: docs/template_extension/unpool_op.hpp [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+//! [fft_op:header]
+#pragma once
+
+#include <ngraph/ngraph.hpp>
+
+namespace TemplateExtension {
+
+class UnpoolOp : public ngraph::op::Op {
+public:
+    static constexpr ngraph::NodeTypeInfo type_info {"MaxPoolGrad", 0};
```
Comment (ilyachur) [inline docs/template_extension/unpool_op.hpp:14]: Please use `NGRAPH_RTTI_DECLARATION` macro

**Review by ilyachur — COMMENTED** [review]
We don't need this operation for our documentation. It looks like a hack if we add `template_extension` to python package.

File: inference-engine/ie_bridges/python/wheel/setup.py (hunk 1 above)
Comment (dkurt) [reply inference-engine/ie_bridges/python/wheel/setup.py:91]: Our goal is to enable the demo https://github.com/openvinotoolkit/openvino_notebooks/pull/46. At this moment there is no other alternative to make it work but MO+nGraph+plugins enablement

**Comment by ilya-lavrenov** [comment]
template extensions cannot be a part of the package.

**Comment by dkurt** [comment]
Is there alternative? There are other requests for easy way for CPU only layers proposals.

**Comment by ilya-lavrenov** [comment]
Create your own extension based on template one

**Comment by dkurt** [comment]
@ilya-lavrenov, Yeah, we tried it for multiple times but nobody wanted to include a build step into their pipeline. See https://github.com/openvinotoolkit/open_model_zoo/pull/949 and https://github.com/openvinotoolkit/openvino_notebooks/pull/46#issuecomment-829214074. Actually, we go back to the strategy with own custom distribution (like https://pypi.org/project/openvino-extensions)

**Comment by ryanloney** [comment]
@ilya-lavrenov Your comment "Create your own extension based on template one" - this is exactly what we are trying to avoid. We don't want to maintain a separate SDL and project just for this extension. Please elaborate on your reason for rejecting this.

**Pull request closed without merging**